File: problem 17/solution.cpp
#include <stack>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st; // stack for opening brackets
        unordered_map<char, char> hash = {{')', '('}, {']', '['}, {'}', '{'}};
        
        for (char c : s) {
            if (hash.count(c)) { // if c is a closing bracket
                if (!st.empty() && st.top() == hash[c]) {
                    st.pop(); // pop the matching opening bracket
                } else {
                    return false; // invalid if no match
                }
            } else {
                st.push(c); // push opening brackets
            }
        }
        return st.empty(); // valid if stack is empty
    }
};

File: problem 17/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        # initialize an empty stack and a hash map for matching brackets
        stack = []
        hash = {')': '(', ']': '[', '}': '{'}
        
        # loop through each character in the string
        for char in s:
            if char in hash:
                # if it's a closing bracket, check the stack
                if stack and stack[-1] == hash[char]:
                    stack.pop()  # remove the matching opening bracket
                else:
                    return False  # invalid if no match
            else:
                # push opening brackets onto the stack
                stack.append(char)
        
        # return True if stack is empty, False otherwise
        return not stack
